src/menu/columnar_menu.rs

```diff
@@ -299,10 +299,22 @@ impl ColumnarMenu {
         use_ansi_coloring: bool,
     ) -> String {
         if use_ansi_coloring {
-            let match_len = self.working_details.shortest_base_string.len();
+            let match_len = self
+                .working_details
+                .shortest_base_string
+                .trim_end()
+                .chars()
+                .count();
 
             // Split string so the match text can be styled
-            let (match_str, remaining_str) = suggestion.value.split_at(match_len);
+            let (match_str, remaining_str) = if match_len <= suggestion.value.chars().count() {
```

Comment:
As the `self.working_details.shortest_base_string` succeeded to be generated in `self.update_values()` (the range coming from `Completer::complete_with_base_ranges` did not violate char boundaries which would panic the `SliceIndex`) 
https://github.com/nushell/reedline/blob/46f410bfdce0b64c3d94fdc0271886f30d64b151/src/menu/columnar_menu.rs#L497-L504
 having to do char based indexing, indicates to me that `shortest_base_string` seems to be not a strict prefix of `suggestion.value`. 
 Don't know if that is really the intended semantics to then split here based on length.

Reply:
the ranges from ``Completer::complete_with_base_ranges`` (and pretty much all buffer related stuff) are byte based not character based, to support unicode we need to do char indexing here.

> shortest_base_string seems to be not a strict prefix of suggestion.value

thats actually somewhat true if you use other completion methods (instead of the default starts_with)
if thats the case ``if match_len <= suggestion.value.chars().count()`` will most likely fail and not style the matching string at all.

Reply:
If the byte indxes are wrong it should already fail on the index in L502, if I remember the slice index semantics of `str` correctly.

So this whole section doing the match highlighting either needs to be aware of which matching mode was used, get the correct coordinates to highlight or perform some alignment to check which characters in the query match to the picked Suggestion.

Reply:
> Im pretty sure the indexing in L502 wont fail because the range is directly coming from the completer (doesnt matter which completion mode was used)

The `Completer` implementer can deliver invalid (i.e. not byte-encoding aware or out-of bounds) ranges in its span (or completely override `complete_with_base_ranges`).
But this would panic L502 as it tries to index the buffer `&str`. [Indexing outside the encoded chars/bounds of a `str` is a panic](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=308c9a8363567e8f769f979a338de2c2)
So the culprit for what you a trying to fix is not a malformed span from the `Completer` (which should be fixed at that source first and then hardened here in a way that still informs implementers).

> if you use other completion methods (instead of the default starts_with)
> if thats the case `if match_len <= suggestion.value.chars().count()` will most likely fail and not style the matching string at all.

We shouldn't accept a highlighting system that only stops to highlight mismatching prompted text and suggestion under the circumstance that something else is fishy.

The highlighting should mark the text that is equal/equivalent between the user query and the suggestions.
And the assumption that the user query is a prefix of all suggestions is not true. 

(I don't have the time to do a full archaeology on #730 or before if the `shortest_base_string` is the right proxy to reflect the matching text itself)

Reply:
I think ill close this PR for now and make another one where i will implement your suggestions. (so add proper support for different completion methods and figure out a safer way to get the base ranges)